File: leetcode/easy/0226InvertBinaryTree/invert_tree.cpp
#include "bits/stdc++.h";
using namespace std;
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class IterativeSolution
{
public:
    TreeNode *invertTree(TreeNode *root)
    {
        if (root == nullptr)
            return nullptr;
        queue<TreeNode*> Queue;
        Queue.push(root);
        while (!Queue.empty())
        {
            TreeNode *current = Queue.front();
            Queue.pop();
            TreeNode *temp = current->left;
            current->left = current->right;
            current->right = temp;
            if (current->left != nullptr)
                Queue.push(current->left);
            if (current->right != nullptr)
                Queue.push(current->right);
        }
        return root;
    }
};

class RecursiveSolution
{
public:
    TreeNode *invertTree(TreeNode *root)
    {
        if (root == nullptr)
        {
            return nullptr;
        }
        TreeNode *right = invertTree(root->right);
        TreeNode *left = invertTree(root->left);
        root->left = right;
        root->right = left;
        return root;
    }
};

File: leetcode/easy/0226InvertBinaryTree/golang/invert_tree.go
package golang

// TreeNode Definition
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func invertTreeRecursive(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}

	queue := make([]*TreeNode, 0)
	queue = append(queue, root)

	for len(queue) != 0 {
		current := queue[0]
		queue = queue[1:]
		temp := current.Right
		current.Right = current.Left
		current.Left = temp

		if current.Right != nil {
			queue = append(queue, current.Right)
		}
		if current.Left != nil {
			queue = append(queue, current.Left)
		}

	}
	return root
}

func invertTree(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}
	right := invertTree(root.Right)
	left := invertTree(root.Left)

	root.Right = left
	root.Left = right

	return root
}
